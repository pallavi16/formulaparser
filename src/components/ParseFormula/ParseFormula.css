.calculator-container {
  background-color: var(--calculator-background-color);
  height: var(--calculator-height);
  width: var(--calculator-width);
}


.submit{
	border-color: #2196F3;
  	color: dodgerblue;
  	width: 90px;
    height: 40px;
    border-radius: 10px;

}

.submit:hover{
	background: #2196F3;
  	color: white;
}


	