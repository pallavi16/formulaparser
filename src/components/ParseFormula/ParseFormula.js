import React, { Component } from 'react';
import './ParseFormula.css';
import Axios from 'axios';



class ParseFormula extends Component {

  state = {
     data: [],
     formValues:{},
     isLoading:true,
     result: []
    };
      doArithmetic = this.doArithmetic.bind(this);  
      modifyExpression = this.modifyExpression.bind(this);
      calculateResult = this.calculateResult.bind(this);


componentDidMount() {
  console.log("Data: "+ this.state.data)

  Axios.get('http://localhost:3000/parser') // JSON File Path
      .then(res => {
        
        this.setState({
          data:res.data,
          isLoading: false
          
        });
       
})
.catch(function(error) {
  this.setState({error, isLoading: false})
});
}


 handleChange(event){
    event.preventDefault();
    let formValues= this.state.formValues;
    let name = event.target.name;
    let value = event.target.value;
    formValues[name] = value;

    this.setState({formValues: formValues});

     formValues = JSON.stringify(this.state.formValues);
    console.log("STRINGIFIED VALS: "+ formValues);

  }


handleClick(event){

console.log("inside fetch");
event.preventDefault();


var data = this.state.data;
var values= this.state.formValues;
var result = this.state.result;
var obj;

fetch('./input1.json')
.then(response => {
  alert("success");
    obj = JSON.parse(JSON.stringify(data));
    Object.keys(obj).map(
       function(object,key){
       if(key === 1){ 
      obj[object]["values"] = values;
      obj[object]["result"] = result;
      }

      }
    )

    this.writeJSON(obj);
})
.catch(err => console.error(err))


  }


writeJSON(formData){

  fetch('http://localhost:3000/parser', { //OR 'http://localhost:3000/db' -same result
    method: 'POST',
    body: JSON.stringify(formData),
    headers: new Headers({
        "Accept": 'application/json',
        "Content-type": "application/json"
        //"Content-type": "application/json; raw"
    }),
}).then(res => res.json())
    .then( res =>{
        console.log('Request succeeded with JSON response', res);

    })
    .catch(function (error) {
        console.log('Request failed', error);
    });
}



  modifyExpression(formula){

    const escapeRE = new RegExp(/[A-Za-z]/g);  
    var characters = this.state.formValues;
    var result;
    console.log("this state form values");
    console.log(characters);
    var s = formula;
   
    if(s !== undefined){
    s = s.replace(escapeRE, m => characters[m]);
    }
    console.log(s);

    if(!/[a-zA-Z]/.test(s)){
      
      result = this.calculateResult(s);
     

    return (
        <div>
          <h4>Result:{result}</h4>

        </div>
      )
  
}
}



  doArithmetic(arithmetic){
    
    console.log(arithmetic);
    var arr;
    var operator = "/*+-^";

    for(var i = 0; i < operator.length; i++){
       if(i === 0)
          {
            arr = arithmetic !== undefined ? arithmetic.split(operator[i]):0;
          }
        else 
          {
          arr = arr.toString().split(operator[i]);
          }
  }
    console.log(arr);
    
   var ar = arr.toString().split(',');
   console.log(ar);


  ar = ar.filter(function (a) {
    return !/^\d+$/.test(a);
  });

     console.log("new array: ");
        console.log(ar);

    return (

       <div style={{width: '100%'}}>
      {
        ar.map((array, idx) =>(
            <div key={idx}>
            {
            array.toString().split(',').map(item =>(
             
                <span>
                  <label htmlFor={idx}> {item}:&nbsp;&nbsp;&nbsp;&nbsp; 
                   <input type="number" name={item} id={idx} placeholder= {item} value={this.state.formValues[item] || ''}  
                   onChange={this.handleChange.bind(this)} />
                   </label>  
                  <br/>
                </span>
              
              ))
            }
            </div>
          ))
      }
      </div>
      );
      

  }

  calculateResult(expr){

    var result = this.state.result;
    var chars = expr.toString().split("");

    var n = [], op = [], index = 0, oplast = true;

    n[index] = "";

    // Parse the expression
    for (var c = 0; c < chars.length; c++) {
        if (isNaN(parseInt(chars[c])) && chars[c] !== "." && !oplast) {
            op[index] = chars[c];
            index++;
            n[index] = "";
            oplast = true;
        } else {
            n[index] += chars[c];
            oplast = false;
        }
    }

    // Calculate the expression
    expr = parseFloat(n[0]);
    for (var o = 0; o < op.length; o++) {
        var num = parseFloat(n[o + 1]);
        switch (op[o]) {
            case "+":
                expr = expr + num;
                break;
            case "-":
                expr = expr - num;
                break;
            case "*":
                expr = expr * num;
                break;
            case "/":
                expr = expr / num;
                break;
            default:
                console.log("arithmetic operator type not supported");    
        }
    }
    console.log("Result is: "+ expr);

          result.push(expr);
  
    return expr;
}
  





render() {
    const { isLoading, data} = this.state;

    return (
      <React.Fragment>
      <div className="calculater-container">
        <h2>Formula</h2>
        <div>
          {!isLoading ? (
           
            data.map(post => {
              const { formula, key} = post;
              
              return (
                <div key={key} >
                  <h3>Formula: {formula}</h3>
                  <h5>Inputs:</h5>
                  {this.doArithmetic(formula)}
                   {this.modifyExpression(formula)} 

                  <hr />
                  </div>
              );
            })
          ) : (
            <p>Loading...</p>
          )}
        </div>
      <button onClick={this.handleClick.bind(this)}> ADD
      </button>
      </div>
      </React.Fragment>
    ); 
  }
}


export default ParseFormula;
