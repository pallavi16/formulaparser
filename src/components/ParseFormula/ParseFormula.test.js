import React from 'react';
import { mount, shallow } from 'enzyme';
import { create } from "react-test-renderer";
import ParseFormula from './ParseFormula';
import Axios from 'axios';
import { expect } from 'chai'
import sinon from 'sinon'



describe('ParseFormula', () => {
  let wrapper;
  beforeEach(() => wrapper = shallow(<ParseFormula />));

  it('should render correctly', () => expect(wrapper).toMatchSnapshot());

  it('should render a <div />', () => {
    expect(wrapper.find('div').length).toEqual(2);
  });
});


describe('mounted Component', () => {
  let wrapper;
  beforeEach(() => wrapper = mount(<ParseFormula />));
it('should fetch data', async () => {
      const props = { get: Promise.resolve('data') };
      const shallowWrapper = shallow(<ParseFormula {...props}/>);
      await props.get;
      expect(shallowWrapper.html()).toBe("<div class=\"calculater-container\"><h2>Formula</h2><div><p>Loading...</p></div><button> ADD</button></div>");
   });
});


describe("Formulas component", () => {
  it("shows a list of formulas", async () => {
    const response = {
      data: [{ formula: "a+b+2+x" }, {formula: "a/d*f+5" }]
    };
    Axios.get(response);
    const component = create(<ParseFormula />);
    const instance = component.getInstance();
    await instance.componentDidMount();
  });
});

describe('Formula component', () => {
  it('should call handleClick() when clicked', () => {
    const spy = sinon.spy(ParseFormula.prototype, 'handleClick')
    const wrapper = shallow(<ParseFormula />)
    wrapper.find('div').simulate('click')      
    expect(spy.calledOnce).to.equal(true)
  })
})