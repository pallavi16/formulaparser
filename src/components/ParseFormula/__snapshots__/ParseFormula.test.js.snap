// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParseFormula should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ParseFormula />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <div
        className="calculater-container"
      >
        <h2>
          Formula
        </h2>
        <div>
          <p>
            Loading...
          </p>
        </div>
        <button
          onClick={[Function]}
        >
           ADD
        </button>
      </div>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2>
            Formula
          </h2>,
          <div>
            <p>
              Loading...
            </p>
          </div>,
          <button
            onClick={[Function]}
          >
             ADD
          </button>,
        ],
        "className": "calculater-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Formula",
          },
          "ref": null,
          "rendered": "Formula",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <p>
              Loading...
            </p>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Loading...",
            },
            "ref": null,
            "rendered": "Loading...",
            "type": "p",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": " ADD",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": " ADD",
          "type": "button",
        },
      ],
      "type": "div",
    },
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <div
          className="calculater-container"
        >
          <h2>
            Formula
          </h2>
          <div>
            <p>
              Loading...
            </p>
          </div>
          <button
            onClick={[Function]}
          >
             ADD
          </button>
        </div>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h2>
              Formula
            </h2>,
            <div>
              <p>
                Loading...
              </p>
            </div>,
            <button
              onClick={[Function]}
            >
               ADD
            </button>,
          ],
          "className": "calculater-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Formula",
            },
            "ref": null,
            "rendered": "Formula",
            "type": "h2",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <p>
                Loading...
              </p>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Loading...",
              },
              "ref": null,
              "rendered": "Loading...",
              "type": "p",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": " ADD",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": " ADD",
            "type": "button",
          },
        ],
        "type": "div",
      },
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
